flatten.sh

```diff
@@ -0,0 +1,277 @@
+    # Make link to PR.
+    git log -1 --format=%s "$1" |
+      sed -nr 's|Merge pull request #([0-9]+).*|Reviewed-on: https://github.com/flang-compiler/f18/pull/\1|p'
+}
+
+try_apply_patches_from_merge() {
+    local merge_commit=${1}
+    local mainline=${merge_commit}^1
+    local branch=${merge_commit}^2
+
+    local on_mainline_during_branch=$(git rev-list --count ${branch}..${mainline})
+    local on_branch=$(git rev-list --count ${mainline}..${branch})
+
+    # git rev-list --reverse --no-merges ${mainline}..${branch} > orig_shas.txt
+    # cp orig_shas.txt rebased_shas.txt
+
+    if [[ $on_mainline_during_branch != 0 ]]
+    then
+        # Stuff happened on mainline since the branch forked. The branch needs
+        # rebasing. If the rebase is trivial, and we end up with the same tree
+        # after playing all the commits, then take it. Otherwise need a fallback
+        # strategy.
+
+        git checkout -B temporary-rebase $branch
+        if git rebase --onto $PARENT ${mainline} temporary-rebase
+        then
+            # Successful rebase.
+            ((N_SUCCESSFUL_REBASE++)) || true
+
+            # Now it's an "easy case", with a new $mainline and $branch.
+            mainline=$PARENT
+            branch=$(git rev-parse temporary-rebase)
+
+            git rev-list --reverse ${mainline}..${branch} > rebased_shas.txt
+            # Rather than just taking PARENT=$(git rev-parse temporary-rebase),
+            # Fall through to easy-case handling to get metadata and trailers.
+        else
+            # These merge commits need ${merge_commit}^2 rebasing onto
+            # ${merge_commit}^1, which is non-trivial work.
+            echo $merge_commit >> hard.txt
+            git rebase --abort
+
+            ((N_HARD_MERGES++)) || true
+            ((N_HARD_COMMITS += $on_branch)) || true
+            return 1
+        fi
+    fi
+
+    # Nothing happened on mainline during branch so make the merge equivalent to
+    # a fast-forward.
+    ((N_EASY++)) || true
+
+    # TODO(peterwaller-arm): See below about Original-Commit.
+    # if [ $(wc -l < rebased_shas.txt) != $(wc -l < orig_shas.txt) ];
+    # then
+    #     echo Failed. 
+    #     cat rebased_shas.txt
+    #     echo ...
+    #     cat orig_shas.txt
+    #     exit 2
+    # fi
+
+    for SHA in $(git rev-list --reverse ${mainline}..${branch})
+    # paste -d@ rebased_shas.txt orig_shas.txt | while IFS="@" read -r SHA ORIG_SHA
+    do
+        # TODO(peterwaller-arm): Need to obtain the original commit sha from
+        # before the rebase above, but I couldn't yet get this working due to
+        # some interesting topologies in the original history.
+        ORIG_SHA=$SHA
+        # echo SHA=$SHA
+        # echo ORIG_SHA=$ORIG_SHA
+        ((N_PRESERVED++)) || true
+        {
+            git show --no-patch --format=%B "$SHA"
+            # echo Original-Commit: flang-compiler/flang@${SHA}
+            make_reviewed_on ${merge_commit}
+        } > message.txt
```

Comment:
I have one small suggestion here: It's good to have the link back to the original pull request, but I think it should be easier to tell which PR a commit is from, for example, in the output of `git log --oneline`. We've lost the information that the commits were grouped together in a pull request.

My suggestion is to prefix the title with something like "PR123: " when the commit comes from pull request 123.

Reply:
In the latest flatten.cpp now pushed, I wrote it as `[flang-compiler/f18#NNN]` so that the cross reference will not reference the LLVM repository, but this repository.